In [2]:
# 210. Compute the historical mean or median per grid and hour-of-day bucket.

import pandas as pd
import numpy as np
from sqlalchemy import create_engine

# MySQL connection
engine = create_engine(
    "mysql+pymysql://root:root@localhost:3306/nopis"
)

# Load network activity data
query = """
SELECT
    grid_id,
    time_key,
    total_activity
FROM fact_network_activity
"""

df = pd.read_sql(query, engine)

# Get timestamp information
time_query = """
SELECT
    time_key,
    timestamp
FROM dim_time
"""

time_df = pd.read_sql(time_query, engine)

# Combine activity with timestamp
df = df.merge(time_df, on="time_key", how="left")

# Create hour-of-day bucket
df["hour"] = pd.to_datetime(df["timestamp"]).dt.hour

# Calculate historical median activity
# for each grid and hour-of-day
baseline_df = (
    df.groupby(["grid_id", "hour"])["total_activity"]
      .median()
      .reset_index()
      .rename(columns={"total_activity": "baseline_value"})
)

# Count how many observations contributed to each baseline
bucket_counts = (
    df.groupby(["grid_id", "hour"])
      .size()
      .reset_index(name="bucket_count")
)

# Add counts to baseline
baseline_df = baseline_df.merge(
    bucket_counts,
    on=["grid_id", "hour"],
    how="left"
)

print("Baseline rows:", len(baseline_df))
print("Unique grids:", baseline_df["grid_id"].nunique())
print("Unique hour buckets:", baseline_df["hour"].nunique())

print("\nSample baseline:")
display(baseline_df.head(10))

print("\nBucket count summary:")
print(baseline_df["bucket_count"].describe())

Baseline rows: 237483
Unique grids: 9998
Unique hour buckets: 24

Sample baseline:


,grid_id,hour,baseline_value,bucket_count
0,1,0,39.1093,7
1,1,1,59.8345,7
2,1,2,67.2336,7
3,1,3,72.8649,7
4,1,4,75.6603,7
5,1,5,73.2843,7
6,1,6,73.0264,7
7,1,7,72.4360,7
8,1,8,80.7752,7
9,1,9,80.9131,7



Bucket count summary:
count    237483.000000
mean          6.552915
std           1.220986
min           1.000000
25%           7.000000
50%           7.000000
75%           7.000000
max           7.000000
Name: bucket_count, dtype: float64


In [4]:
# 211. Reuse the NP3 baseline function rather than writing a second one.

def calculate_baseline(
    data,
    group_columns,
    value_column="total_activity",
    statistic="median"
):
    """
    Generalized baseline calculation.

    group_columns:
        Columns used to define comparable historical observations.

    statistic:
        'median' or 'mean'
    """

    if statistic == "median":
        baseline = (
            data.groupby(group_columns)[value_column]
            .median()
            .reset_index()
        )
    elif statistic == "mean":
        baseline = (
            data.groupby(group_columns)[value_column]
            .mean()
            .reset_index()
        )
    else:
        raise ValueError("statistic must be 'median' or 'mean'")

    baseline = baseline.rename(
        columns={value_column: "baseline_value"}
    )

    return baseline


# Test the generalized function using the ML4
# grid + hour-of-day baseline
hourly_baseline = calculate_baseline(
    df,
    group_columns=["grid_id", "hour"],
    value_column="total_activity",
    statistic="median"
)

print("Hourly baseline rows:", len(hourly_baseline))
print("Unique grids:", hourly_baseline["grid_id"].nunique())
print("Unique hours:", hourly_baseline["hour"].nunique())

display(hourly_baseline.head(10))

Hourly baseline rows: 237483
Unique grids: 9998
Unique hours: 24


,grid_id,hour,baseline_value
0,1,0,39.1093
1,1,1,59.8345
2,1,2,67.2336
3,1,3,72.8649
4,1,4,75.6603
5,1,5,73.2843
6,1,6,73.0264
7,1,7,72.4360
8,1,8,80.7752
9,1,9,80.9131


In [3]:
# Check which grids exist in dim_grid but have no activity data

grid_query = """
SELECT grid_id
FROM dim_grid
"""

grid_df = pd.read_sql(grid_query, engine)

activity_grids = set(df["grid_id"].unique())
all_grids = set(grid_df["grid_id"].unique())

missing_grids = sorted(all_grids - activity_grids)

print("Total grids in dim_grid:", len(all_grids))
print("Grids in fact_network_activity:", len(activity_grids))
print("Missing grids:", len(missing_grids))
print("Missing grid IDs:", missing_grids)

Total grids in dim_grid: 10000
Grids in fact_network_activity: 9998
Missing grids: 2
Missing grid IDs: [np.int64(5239), np.int64(5339)]


In [5]:
# 212. Compute the deviation from the baseline.

# Add hour-of-day to the activity data
df["hour"] = pd.to_datetime(df["timestamp"]).dt.hour

# Merge the historical baseline with the activity data
anomaly_df = df.merge(
    hourly_baseline,
    on=["grid_id", "hour"],
    how="left"
)

# Calculate deviation
anomaly_df["deviation"] = (
    anomaly_df["total_activity"] -
    anomaly_df["baseline_value"]
)

print("Rows:", len(anomaly_df))
print("Missing baselines:", anomaly_df["baseline_value"].isna().sum())

print("\nSample:")
display(
    anomaly_df[
        ["grid_id", "timestamp", "hour",
         "total_activity", "baseline_value", "deviation"]
    ].head(10)
)

Rows: 1556206
Missing baselines: 0

Sample:


,grid_id,timestamp,hour,total_activity,baseline_value,deviation
0,1,2013-10-31 18:30:00,18,61.6037,57.20910,4.39460
1,1,2013-10-31 19:30:00,19,46.1967,45.99925,0.19745
2,1,2013-10-31 20:30:00,20,42.0324,40.71960,1.31280
3,1,2013-10-31 21:30:00,21,35.0705,35.07050,0.00000
4,1,2013-10-31 22:30:00,22,32.2687,32.35365,-0.08495
5,1,2013-10-31 23:30:00,23,35.2842,31.88105,3.40315
6,1,2013-11-01 00:30:00,0,35.9625,39.10930,-3.14680
7,1,2013-11-01 01:30:00,1,44.4976,59.83450,-15.33690
8,1,2013-11-01 02:30:00,2,65.8393,67.23360,-1.39430
9,1,2013-11-01 03:30:00,3,82.0947,72.86490,9.22980


In [6]:
# 213. Define an anomaly score using a simple standardized or percentage deviation.

# Calculate percentage deviation from the historical baseline
anomaly_df["anomaly_score"] = (
    anomaly_df["deviation"] /
    anomaly_df["baseline_value"]
) * 100

# Check the results
print("Missing anomaly scores:", anomaly_df["anomaly_score"].isna().sum())
print("Infinite anomaly scores:", np.isinf(anomaly_df["anomaly_score"]).sum())

print("\nAnomaly score summary:")
print(anomaly_df["anomaly_score"].describe())

print("\nSample:")
display(
    anomaly_df[
        ["grid_id", "timestamp",
         "total_activity", "baseline_value",
         "deviation", "anomaly_score"]
    ].head(10)
)

Missing anomaly scores: 0
Infinite anomaly scores: 0

Anomaly score summary:
count    1.556206e+06
mean    -1.884293e+00
std      2.867246e+01
min     -9.991202e+01
25%     -1.172492e+01
50%      0.000000e+00
75%      7.898409e+00
max      5.537474e+03
Name: anomaly_score, dtype: float64

Sample:


,grid_id,timestamp,total_activity,baseline_value,deviation,anomaly_score
0,1,2013-10-31 18:30:00,61.6037,57.20910,4.39460,7.681645
1,1,2013-10-31 19:30:00,46.1967,45.99925,0.19745,0.429246
2,1,2013-10-31 20:30:00,42.0324,40.71960,1.31280,3.224000
3,1,2013-10-31 21:30:00,35.0705,35.07050,0.00000,0.000000
4,1,2013-10-31 22:30:00,32.2687,32.35365,-0.08495,-0.262567
5,1,2013-10-31 23:30:00,35.2842,31.88105,3.40315,10.674523
6,1,2013-11-01 00:30:00,35.9625,39.10930,-3.14680,-8.046168
7,1,2013-11-01 01:30:00,44.4976,59.83450,-15.33690,-25.632202
8,1,2013-11-01 02:30:00,65.8393,67.23360,-1.39430,-2.073814
9,1,2013-11-01 03:30:00,82.0947,72.86490,9.22980,12.667004


In [7]:
# 214. Flag both unusually high and unusually low activity, and keep the direction as a field.

ANOMALY_THRESHOLD = 50

# Create anomaly flag
anomaly_df["anomaly_flag"] = (
    anomaly_df["anomaly_score"].abs() >= ANOMALY_THRESHOLD
)

# Record the direction
anomaly_df["direction"] = np.select(
    [
        anomaly_df["anomaly_score"] >= ANOMALY_THRESHOLD,
        anomaly_df["anomaly_score"] <= -ANOMALY_THRESHOLD
    ],
    [
        "HIGH",
        "LOW"
    ],
    default="NORMAL"
)

# Check counts
print("Direction counts:")
print(anomaly_df["direction"].value_counts())

print("\nAnomaly flag counts:")
print(anomaly_df["anomaly_flag"].value_counts())

print("\nSample HIGH anomalies:")
display(
    anomaly_df[
        anomaly_df["direction"] == "HIGH"
    ][
        ["grid_id", "timestamp", "total_activity",
         "baseline_value", "anomaly_score", "direction"]
    ].head(5)
)

print("\nSample LOW anomalies:")
display(
    anomaly_df[
        anomaly_df["direction"] == "LOW"
    ][
        ["grid_id", "timestamp", "total_activity",
         "baseline_value", "anomaly_score", "direction"]
    ].head(5)
)

Direction counts:
direction
NORMAL    1465235
LOW         60279
HIGH        30692
Name: count, dtype: int64

Anomaly flag counts:
anomaly_flag
False    1465235
True       90971
Name: count, dtype: int64

Sample HIGH anomalies:


,grid_id,timestamp,total_activity,baseline_value,anomaly_score,direction
109,1,2013-11-05 16:30:00,148.6847,85.4263,74.050263,HIGH
264,2,2013-11-05 16:30:00,149.0953,85.7914,73.788165,HIGH
416,3,2013-11-05 16:30:00,149.5322,86.1799,73.511689,HIGH
568,4,2013-11-05 16:30:00,147.4957,84.3689,74.822358,HIGH
723,5,2013-11-05 16:30:00,132.0592,76.4992,72.628210,HIGH



Sample LOW anomalies:


,grid_id,timestamp,total_activity,baseline_value,anomaly_score,direction
5940,41,2013-11-01 09:30:00,58.8179,122.0267,-51.799155,LOW
5985,41,2013-11-03 07:30:00,52.7256,108.7984,-51.538258,LOW
5987,41,2013-11-03 09:30:00,60.4573,122.0267,-50.455679,LOW
6119,42,2013-11-02 12:30:00,56.7336,113.5018,-50.015242,LOW
6229,43,2013-11-01 02:30:00,47.1781,121.2274,-61.082973,LOW


In [10]:
# 215. Compare the anomaly flag against the ML3 classifier output and the NP3 rule alerts, and explain the disagreements.

# ---------------------------------------------------------
# Recreate the ML3 Logistic Regression model
# ---------------------------------------------------------

from sklearn.linear_model import LogisticRegression

ML3_THRESHOLD = 1154.5988

# Load ML3 features and next-hour activity
ml3_query = """
SELECT
    grid_id,
    feature_timestamp AS timestamp,
    avg_activity,
    activity_growth,
    active_hours,
    peak_ratio,
    variability,
    internet_share
FROM network_feature_table
"""

ml3_data = pd.read_sql(ml3_query, engine)

activity_query = """
SELECT
    grid_id,
    time_key,
    total_activity
FROM fact_network_activity
"""

activity_data = pd.read_sql(activity_query, engine)

time_query = """
SELECT
    time_key,
    timestamp
FROM dim_time
"""

time_data = pd.read_sql(time_query, engine)

activity_data = activity_data.merge(
    time_data,
    on="time_key",
    how="left"
)

activity_data["timestamp"] = pd.to_datetime(
    activity_data["timestamp"]
)

ml3_data["timestamp"] = pd.to_datetime(
    ml3_data["timestamp"]
)

# Create next-hour target
activity_data = activity_data.sort_values(
    ["grid_id", "timestamp"]
)

activity_data["target"] = (
    activity_data
    .groupby("grid_id")["total_activity"]
    .shift(-1)
    > ML3_THRESHOLD
).astype(float)

# Merge target with ML3 features
ml3_data = ml3_data.merge(
    activity_data[
        ["grid_id", "timestamp", "target"]
    ],
    on=["grid_id", "timestamp"],
    how="left"
)

# Remove rows without complete features or target
feature_columns = [
    "avg_activity",
    "activity_growth",
    "active_hours",
    "peak_ratio",
    "variability",
    "internet_share"
]

ml3_data = ml3_data.dropna(
    subset=feature_columns + ["target"]
)

ml3_data["target"] = ml3_data["target"].astype(int)

# Chronological 80/20 split
ml3_data = ml3_data.sort_values("timestamp")

split_index = int(len(ml3_data) * 0.80)

train_data = ml3_data.iloc[:split_index]
test_data = ml3_data.iloc[split_index:]

# Train the same ML3 model
model = LogisticRegression(max_iter=1000)

model.fit(
    train_data[feature_columns],
    train_data["target"]
)

print("ML3 model recreated successfully.")
print("Training rows:", len(train_data))
print("Test rows:", len(test_data))

ML3 model recreated successfully.
Training rows: 869040
Test rows: 217260


In [11]:
# 215. Compare the anomaly flag against the ML3 classifier output and the NP3 rule alerts, and explain the disagreements.

# ---------------------------------------------------------
# Load ML3 features
# ---------------------------------------------------------

feature_query = """
SELECT
    grid_id,
    feature_timestamp AS timestamp,
    avg_activity,
    activity_growth,
    active_hours,
    peak_ratio,
    variability,
    internet_share
FROM network_feature_table
"""

ml3_features = pd.read_sql(feature_query, engine)

ml3_features["timestamp"] = pd.to_datetime(
    ml3_features["timestamp"]
)

# ---------------------------------------------------------
# Merge ML3 features with anomaly results
# ---------------------------------------------------------

comparison_df = anomaly_df.merge(
    ml3_features,
    on=["grid_id", "timestamp"],
    how="left"
)

# ---------------------------------------------------------
# Generate ML3 predictions
# ---------------------------------------------------------

feature_columns = [
    "avg_activity",
    "activity_growth",
    "active_hours",
    "peak_ratio",
    "variability",
    "internet_share"
]

valid_ml3 = comparison_df[feature_columns].notna().all(axis=1)

comparison_df["ml_prediction"] = 0

comparison_df.loc[valid_ml3, "ml_prediction"] = (
    model.predict(
        comparison_df.loc[valid_ml3, feature_columns]
    )
)

comparison_df["ml_prediction"] = (
    comparison_df["ml_prediction"].astype(int)
)

# ---------------------------------------------------------
# NP3 HIGH_ACTIVITY rule
# ---------------------------------------------------------

NP3_HIGH_ACTIVITY_THRESHOLD = 1.50

comparison_df["np3_alert"] = (
    comparison_df["total_activity"] >=
    comparison_df["baseline_value"] *
    NP3_HIGH_ACTIVITY_THRESHOLD
).astype(int)

# ---------------------------------------------------------
# Convert anomaly flag to 0/1
# ---------------------------------------------------------

comparison_df["anomaly_flag_int"] = (
    comparison_df["anomaly_flag"].astype(int)
)

# ---------------------------------------------------------
# Three-way comparison
# ---------------------------------------------------------

comparison_df["comparison"] = np.select(
    [
        (
            (comparison_df["anomaly_flag_int"] == 1) &
            (comparison_df["ml_prediction"] == 1) &
            (comparison_df["np3_alert"] == 1)
        ),

        (
            (comparison_df["anomaly_flag_int"] == 1) &
            (comparison_df["ml_prediction"] == 1) &
            (comparison_df["np3_alert"] == 0)
        ),

        (
            (comparison_df["anomaly_flag_int"] == 1) &
            (comparison_df["ml_prediction"] == 0) &
            (comparison_df["np3_alert"] == 1)
        ),

        (
            (comparison_df["anomaly_flag_int"] == 0) &
            (comparison_df["ml_prediction"] == 1) &
            (comparison_df["np3_alert"] == 1)
        )
    ],
    [
        "ALL_THREE",
        "ANOMALY_ML_ONLY",
        "ANOMALY_NP3_ONLY",
        "ML_NP3_ONLY"
    ],
    default="OTHER"
)

# ---------------------------------------------------------
# Results
# ---------------------------------------------------------

print("Three-way comparison:")
print(comparison_df["comparison"].value_counts())

print("\nTotal rows:", len(comparison_df))

print(
    "\nRows with valid ML3 features:",
    valid_ml3.sum()
)

print("\nComparison sample:")

display(
    comparison_df[
        [
            "grid_id",
            "timestamp",
            "total_activity",
            "baseline_value",
            "anomaly_score",
            "direction",
            "anomaly_flag",
            "ml_prediction",
            "np3_alert",
            "comparison"
        ]
    ].head(20)
)

Three-way comparison:
comparison
OTHER               1523200
ANOMALY_NP3_ONLY      29237
ANOMALY_ML_ONLY        2314
ALL_THREE              1455
Name: count, dtype: int64

Total rows: 1556206

Rows with valid ML3 features: 1086300

Comparison sample:


,grid_id,timestamp,total_activity,baseline_value,anomaly_score,direction,anomaly_flag,ml_prediction,np3_alert,comparison
0,1,2013-10-31 18:30:00,61.6037,57.20910,7.681645,NORMAL,False,0,0,OTHER
1,1,2013-10-31 19:30:00,46.1967,45.99925,0.429246,NORMAL,False,0,0,OTHER
2,1,2013-10-31 20:30:00,42.0324,40.71960,3.224000,NORMAL,False,0,0,OTHER
3,1,2013-10-31 21:30:00,35.0705,35.07050,0.000000,NORMAL,False,0,0,OTHER
4,1,2013-10-31 22:30:00,32.2687,32.35365,-0.262567,NORMAL,False,0,0,OTHER
5,1,2013-10-31 23:30:00,35.2842,31.88105,10.674523,NORMAL,False,0,0,OTHER
6,1,2013-11-01 00:30:00,35.9625,39.10930,-8.046168,NORMAL,False,0,0,OTHER
7,1,2013-11-01 01:30:00,44.4976,59.83450,-25.632202,NORMAL,False,0,0,OTHER
8,1,2013-11-01 02:30:00,65.8393,67.23360,-2.073814,NORMAL,False,0,0,OTHER
9,1,2013-11-01 03:30:00,82.0947,72.86490,12.667004,NORMAL,False,0,0,OTHER


In [12]:
# 215. Compare the anomaly flag against the ML3 classifier output and the NP3 rule alerts, and explain the disagreements.

# Create a clear label for every combination of the three mechanisms

comparison_df["three_way_result"] = (
    "Anomaly=" + comparison_df["anomaly_flag_int"].astype(str)
    + ", ML3=" + comparison_df["ml_prediction"].astype(str)
    + ", NP3=" + comparison_df["np3_alert"].astype(str)
)

print("All three-way combinations:")
print(comparison_df["three_way_result"].value_counts())

print("\nTotal rows:", len(comparison_df))

# Show only cases where at least one mechanism disagrees
disagreement_df = comparison_df[
    comparison_df["three_way_result"].isin([
        "Anomaly=1, ML3=1, NP3=0",
        "Anomaly=1, ML3=0, NP3=1",
        "Anomaly=0, ML3=1, NP3=1",
        "Anomaly=1, ML3=0, NP3=0",
        "Anomaly=0, ML3=1, NP3=0",
        "Anomaly=0, ML3=0, NP3=1"
    ])
].copy()

print("\nDisagreement cases:", len(disagreement_df))

display(
    disagreement_df[
        [
            "grid_id",
            "timestamp",
            "total_activity",
            "baseline_value",
            "anomaly_score",
            "direction",
            "anomaly_flag",
            "ml_prediction",
            "np3_alert",
            "three_way_result"
        ]
    ].head(20)
)

All three-way combinations:
three_way_result
Anomaly=0, ML3=0, NP3=0    1362400
Anomaly=0, ML3=1, NP3=0     102835
Anomaly=1, ML3=0, NP3=0      57965
Anomaly=1, ML3=0, NP3=1      29237
Anomaly=1, ML3=1, NP3=0       2314
Anomaly=1, ML3=1, NP3=1       1455
Name: count, dtype: int64

Total rows: 1556206

Disagreement cases: 192351


,grid_id,timestamp,total_activity,baseline_value,anomaly_score,direction,anomaly_flag,ml_prediction,np3_alert,three_way_result
109,1,2013-11-05 16:30:00,148.6847,85.4263,74.050263,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
264,2,2013-11-05 16:30:00,149.0953,85.7914,73.788165,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
416,3,2013-11-05 16:30:00,149.5322,86.1799,73.511689,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
568,4,2013-11-05 16:30:00,147.4957,84.3689,74.822358,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
723,5,2013-11-05 16:30:00,132.0592,76.4992,72.628210,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
875,6,2013-11-05 16:30:00,149.5322,86.1799,73.511689,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
1024,7,2013-11-05 16:30:00,149.5322,86.1799,73.511689,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
1173,8,2013-11-05 16:30:00,149.5322,86.1799,73.511689,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
1322,9,2013-11-05 16:30:00,149.5322,86.1799,73.511689,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"
1992,14,2013-11-02 05:30:00,27.9963,18.3868,52.263037,HIGH,True,0,1,"Anomaly=1, ML3=0, NP3=1"


One genuine disagreement

For example, Grid 1 at 2013-11-05 16:30:

Current activity = 148.68
Baseline         = 85.43
Anomaly score    = +74.05%

So the anomaly detector flags HIGH.

At the same time, NP3 also flags it, because:

148.68 / 85.43 = 1.74

which is greater than the NP3 threshold of 1.50.

But ML3 does not flag it.

This is a legitimate disagreement because the three mechanisms use different logic:

Anomaly: compares activity with historical hour-of-day behaviour.
NP3: uses a fixed 1.5× baseline rule.
ML3: combines six learned features to predict the next-hour high-activity target.

So disagreement is actually useful—it tells the operations team that the mechanisms are looking at different signals.

In [13]:
# 216. Store both the score and a human-readable reason.

# Create a human-readable explanation
def create_anomaly_reason(row):
    score = row["anomaly_score"]
    baseline = row["baseline_value"]
    current = row["total_activity"]

    if row["direction"] == "HIGH":
        return (
            f"Activity is {abs(score):.1f}% above the historical "
            f"baseline ({current:.2f} vs {baseline:.2f})."
        )

    elif row["direction"] == "LOW":
        return (
            f"Activity is {abs(score):.1f}% below the historical "
            f"baseline ({current:.2f} vs {baseline:.2f})."
        )

    else:
        return (
            f"Activity is within the normal range "
            f"({current:.2f} vs baseline {baseline:.2f})."
        )


comparison_df["reason"] = comparison_df.apply(
    create_anomaly_reason,
    axis=1
)

# Create the final anomaly score table
network_anomaly_scores = comparison_df[
    [
        "grid_id",
        "timestamp",
        "total_activity",
        "baseline_value",
        "deviation",
        "anomaly_score",
        "direction",
        "anomaly_flag",
        "reason"
    ]
].copy()

# Rename timestamp to match the requested output naming
network_anomaly_scores = network_anomaly_scores.rename(
    columns={
        "timestamp": "feature_timestamp",
        "total_activity": "current_value"
    }
)

print("Rows:", len(network_anomaly_scores))
print("Columns:", list(network_anomaly_scores.columns))

print("\nSample HIGH anomaly reasons:")
display(
    network_anomaly_scores[
        network_anomaly_scores["direction"] == "HIGH"
    ].head(5)
)

print("\nSample LOW anomaly reasons:")
display(
    network_anomaly_scores[
        network_anomaly_scores["direction"] == "LOW"
    ].head(5)
)

print("\nSample NORMAL reasons:")
display(
    network_anomaly_scores[
        network_anomaly_scores["direction"] == "NORMAL"
    ].head(5)
)

Rows: 1556206
Columns: ['grid_id', 'feature_timestamp', 'current_value', 'baseline_value', 'deviation', 'anomaly_score', 'direction', 'anomaly_flag', 'reason']

Sample HIGH anomaly reasons:


,grid_id,feature_timestamp,current_value,baseline_value,deviation,anomaly_score,direction,anomaly_flag,reason
109,1,2013-11-05 16:30:00,148.6847,85.4263,63.2584,74.050263,HIGH,True,Activity is 74.1% above the historical baselin...
264,2,2013-11-05 16:30:00,149.0953,85.7914,63.3039,73.788165,HIGH,True,Activity is 73.8% above the historical baselin...
416,3,2013-11-05 16:30:00,149.5322,86.1799,63.3523,73.511689,HIGH,True,Activity is 73.5% above the historical baselin...
568,4,2013-11-05 16:30:00,147.4957,84.3689,63.1268,74.822358,HIGH,True,Activity is 74.8% above the historical baselin...
723,5,2013-11-05 16:30:00,132.0592,76.4992,55.5600,72.628210,HIGH,True,Activity is 72.6% above the historical baselin...



Sample LOW anomaly reasons:


,grid_id,feature_timestamp,current_value,baseline_value,deviation,anomaly_score,direction,anomaly_flag,reason
5940,41,2013-11-01 09:30:00,58.8179,122.0267,-63.2088,-51.799155,LOW,True,Activity is 51.8% below the historical baselin...
5985,41,2013-11-03 07:30:00,52.7256,108.7984,-56.0728,-51.538258,LOW,True,Activity is 51.5% below the historical baselin...
5987,41,2013-11-03 09:30:00,60.4573,122.0267,-61.5694,-50.455679,LOW,True,Activity is 50.5% below the historical baselin...
6119,42,2013-11-02 12:30:00,56.7336,113.5018,-56.7682,-50.015242,LOW,True,Activity is 50.0% below the historical baselin...
6229,43,2013-11-01 02:30:00,47.1781,121.2274,-74.0493,-61.082973,LOW,True,Activity is 61.1% below the historical baselin...



Sample NORMAL reasons:


,grid_id,feature_timestamp,current_value,baseline_value,deviation,anomaly_score,direction,anomaly_flag,reason
0,1,2013-10-31 18:30:00,61.6037,57.20910,4.39460,7.681645,NORMAL,False,Activity is within the normal range (61.60 vs ...
1,1,2013-10-31 19:30:00,46.1967,45.99925,0.19745,0.429246,NORMAL,False,Activity is within the normal range (46.20 vs ...
2,1,2013-10-31 20:30:00,42.0324,40.71960,1.31280,3.224000,NORMAL,False,Activity is within the normal range (42.03 vs ...
3,1,2013-10-31 21:30:00,35.0705,35.07050,0.00000,0.000000,NORMAL,False,Activity is within the normal range (35.07 vs ...
4,1,2013-10-31 22:30:00,32.2687,32.35365,-0.08495,-0.262567,NORMAL,False,Activity is within the normal range (32.27 vs ...


In [15]:
# Store network anomaly scores in MySQL

from sqlalchemy import text

# Create the table
create_table_query = """
CREATE TABLE IF NOT EXISTS network_anomaly_scores (
    grid_id INT,
    feature_timestamp DATETIME,
    current_value DOUBLE,
    baseline_value DOUBLE,
    deviation DOUBLE,
    anomaly_score DOUBLE,
    direction VARCHAR(10),
    anomaly_flag BOOLEAN,
    reason TEXT
)
"""

with engine.connect() as conn:
    conn.execute(text(create_table_query))
    conn.commit()

# Clear existing data so rerunning the notebook
# does not create duplicate records
with engine.connect() as conn:
    conn.execute(text("TRUNCATE TABLE network_anomaly_scores"))
    conn.commit()

# Store the results
network_anomaly_scores.to_sql(
    "network_anomaly_scores",
    con=engine,
    if_exists="append",
    index=False,
    chunksize=10000,
    method="multi"
)

print("network_anomaly_scores stored successfully.")

network_anomaly_scores stored successfully.


In [16]:
# Final ML4 verification

from sqlalchemy import text

# ---------------------------------------------------------
# 1. Check stored row count
# ---------------------------------------------------------

row_count = pd.read_sql(
    "SELECT COUNT(*) AS row_count FROM network_anomaly_scores",
    engine
)

print("1. Stored rows:")
print(row_count)


# ---------------------------------------------------------
# 2. Check HIGH / LOW / NORMAL anomalies
# ---------------------------------------------------------

direction_counts = pd.read_sql(
    """
    SELECT
        direction,
        COUNT(*) AS count
    FROM network_anomaly_scores
    GROUP BY direction
    ORDER BY count DESC
    """,
    engine
)

print("\n2. Direction counts:")
display(direction_counts)


# ---------------------------------------------------------
# 3. Check anomaly flag
# ---------------------------------------------------------

flag_counts = pd.read_sql(
    """
    SELECT
        anomaly_flag,
        COUNT(*) AS count
    FROM network_anomaly_scores
    GROUP BY anomaly_flag
    """,
    engine
)

print("\n3. Anomaly flag counts:")
display(flag_counts)


# ---------------------------------------------------------
# 4. Check baseline bucket counts
# ---------------------------------------------------------

bucket_check = pd.read_sql(
    """
    SELECT
        COUNT(*) AS baseline_rows,
        COUNT(DISTINCT grid_id) AS unique_grids,
        COUNT(DISTINCT HOUR(feature_timestamp)) AS hour_buckets
    FROM network_anomaly_scores
    """,
    engine
)

print("\n4. Baseline coverage:")
display(bucket_check)


# ---------------------------------------------------------
# 5. Check human-readable reasons
# ---------------------------------------------------------

reason_check = pd.read_sql(
    """
    SELECT
        grid_id,
        feature_timestamp,
        anomaly_score,
        direction,
        reason
    FROM network_anomaly_scores
    WHERE direction IN ('HIGH', 'LOW')
    LIMIT 10
    """,
    engine
)

print("\n5. Sample explanations:")
display(reason_check)


# ---------------------------------------------------------
# 6. Final acceptance checks
# ---------------------------------------------------------

stored_rows = row_count.iloc[0]["row_count"]

has_high = (
    "HIGH" in direction_counts["direction"].values
)

has_low = (
    "LOW" in direction_counts["direction"].values
)

has_reasons = (
    reason_check["reason"].notna().all()
)

print("\n========== ML4 ACCEPTANCE CHECK ==========")

print(
    "PASS: Rows stored"
    if stored_rows > 0
    else "FAIL: No rows stored"
)

print(
    "PASS: HIGH anomalies produced"
    if has_high
    else "FAIL: No HIGH anomalies"
)

print(
    "PASS: LOW anomalies produced"
    if has_low
    else "FAIL: No LOW anomalies"
)

print(
    "PASS: Human-readable reasons stored"
    if has_reasons
    else "FAIL: Missing reasons"
)

print("\nML4 verification complete.")

1. Stored rows:
   row_count
0    1556206

2. Direction counts:


,direction,count
0,NORMAL,1465235
1,LOW,60279
2,HIGH,30692



3. Anomaly flag counts:


,anomaly_flag,count
0,0,1465235
1,1,90971



4. Baseline coverage:


,baseline_rows,unique_grids,hour_buckets
0,1556206,9998,24



5. Sample explanations:


,grid_id,feature_timestamp,anomaly_score,direction,reason
0,1,2013-11-05 16:30:00,74.050263,HIGH,Activity is 74.1% above the historical baselin...
1,2,2013-11-05 16:30:00,73.788165,HIGH,Activity is 73.8% above the historical baselin...
2,3,2013-11-05 16:30:00,73.511689,HIGH,Activity is 73.5% above the historical baselin...
3,4,2013-11-05 16:30:00,74.822358,HIGH,Activity is 74.8% above the historical baselin...
4,5,2013-11-05 16:30:00,72.628210,HIGH,Activity is 72.6% above the historical baselin...
5,6,2013-11-05 16:30:00,73.511689,HIGH,Activity is 73.5% above the historical baselin...
6,7,2013-11-05 16:30:00,73.511689,HIGH,Activity is 73.5% above the historical baselin...
7,8,2013-11-05 16:30:00,73.511689,HIGH,Activity is 73.5% above the historical baselin...
8,9,2013-11-05 16:30:00,73.511689,HIGH,Activity is 73.5% above the historical baselin...
9,14,2013-11-02 05:30:00,52.263037,HIGH,Activity is 52.3% above the historical baselin...



========== ML4 ACCEPTANCE CHECK ==========
PASS: Rows stored
PASS: HIGH anomalies produced
PASS: LOW anomalies produced
PASS: Human-readable reasons stored

ML4 verification complete.
